[Toponaming] create ElementMap class

Pesc0 opened the pull request:
Thank you for creating a pull request to contribute to FreeCAD! Place an "X" in between the brackets below to "check off" to confirm that you have satisfied the requirement, or ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10) if there is something you don't understand.

- [x]  Your Pull Request meets the requirements outlined in section 5 of [CONTRIBUTING.md](https://github.com/FreeCAD/FreeCAD/blob/master/CONTRIBUTING.md) for a Valid PR

Please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [1.0 Changelog Forum Thread](https://forum.freecad.org/viewtopic.php?f=10&t=69438).

---

fixes #9115

Pesc0 commented:
Uhm... I think I need some help
```
src/Base/Console.h:355:28: error: ‘_s_fclvl’ was not declared in this scope
  355 | #   define FC_LOG_INSTANCE _s_fclvl
      |                            ^~~~~~~~
```

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9175"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_9175/pipeline.svg"/></a> for feature branch [PR_9175](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9175). Pipeline [826781146](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/826781146) was triggered at [50e15c8](https://github.com/FreeCAD/FreeCAD/pull/9175/commits/50e15c83e32897d1b5072508796d6f65c04b1cf2). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9175"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_9175/pipeline.svg"/></a> for feature branch [PR_9175](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9175). Pipeline [828829178](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/828829178) was triggered at [b3e6712](https://github.com/FreeCAD/FreeCAD/pull/9175/commits/b3e67120015992e26dfc023018b62fb7ec2c315d). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

chennes commented:
If you add `[skip ci]` to your commit message it will not run CI.

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9175"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_9175/pipeline.svg"/></a> for feature branch [PR_9175](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9175). Pipeline [828953481](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/828953481) was triggered at [dcb5b4f](https://github.com/FreeCAD/FreeCAD/pull/9175/commits/dcb5b4f32fa58d5459d14712665cf6da76b5fbb0). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9175"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_9175/pipeline.svg"/></a> for feature branch [PR_9175](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9175). Pipeline [829000923](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/829000923) was triggered at [bff603f](https://github.com/FreeCAD/FreeCAD/pull/9175/commits/bff603f469af2d7e9d89511b5468164856ab569c). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9175"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_9175/pipeline.svg"/></a> for feature branch [PR_9175](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9175). Pipeline [836875996](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/836875996) was triggered at [11b5da3](https://github.com/FreeCAD/FreeCAD/pull/9175/commits/11b5da36f68d20038f3f4e1ddbf595364ea42152). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9175"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_9175/pipeline.svg"/></a> for feature branch [PR_9175](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9175). Pipeline [838453245](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/838453245) was triggered at [b063c58](https://github.com/FreeCAD/FreeCAD/pull/9175/commits/b063c5814f1749de794f056f5d8178ff1486b8d4). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9175"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_9175/pipeline.svg"/></a> for feature branch [PR_9175](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9175). Pipeline [838898376](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/838898376) was triggered at [6d22e3e](https://github.com/FreeCAD/FreeCAD/pull/9175/commits/6d22e3e7477557b36e5d122f74e05a8a8a7b6e46). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9175"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_9175/pipeline.svg"/></a> for feature branch [PR_9175](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9175). Pipeline [842282405](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/842282405) was triggered at [7fda894](https://github.com/FreeCAD/FreeCAD/pull/9175/commits/7fda89499b87f6ae4a7c41101c0d2baa73de3b3e). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

adrianinsaval commented on .vscode/c_cpp_properties.json:
```diff
@@ -1,12 +1,16 @@
+        }
+    ],
+    "version": 4
+}
```
I don't think you should touch these files unless you know what youa re doing and actually want to change the default for everyone

Pesc0 replied:
Afaik this just adds the two paths so that vscode can find the proper include headers (on linux of course). I've intentionally named the commit [vscode stuff: drop or merge?](https://github.com/FreeCAD/FreeCAD/pull/9175/commits/cf4651588916d857454ca017a1a4cc5f50703a37), so I'm perfectly fine reverting these changes before merge

AjinkyaDahale replied:
@adrianinsaval we should probably remove this folder itself and add it to `.gitignore`

Pesc0 replied:
It was added recently by #9098

adrianinsaval replied:
adding to gitignore would be a good idea, when you actually want to make changes there you can still explicitly add them to the commit. Here that commit should be dropped IMO

Pesc0 replied:
I agree the commit is not really part of this PR. Still these changes made vscode happy on both of my machines (arch and fedora). Not trying to force this or anything, its just my opinion since this has been useful to me. Unless someone is favourable on keeping this I will drop it before merge.

chennes replied:
It probably doesn't make sense to have the include path changes in there, since they are system-specific. The rest of the changes are fine (though I agree that we should add the files to .gitignore).

Pesc0 commented:
rebased on master

AjinkyaDahale commented on src/App/ComplexGeoData.h:
```diff
@@ -166,20 +166,12 @@ class AppExport ComplexGeoData: public Base::Persistence, public Base::Handled
+     * @return Returns the name stripped with ELEMENT_MAP_PREFIX, or 0 if not
      * start with the prefix
      */
     static const char *isMappedElement(const char *name);
```
If we move `isMappedElement`, `newElementName` and `oldElementName` to `ElementMap` or a smaller header, `SketchObject` is no longer depenedent on this header and thus lowering overhead.

Pesc0 commented:
Thanks for the review. I'd argue it's best to get elementmap working for now, this stuff can be moved quickly later in the complexgeodata card. Opinions (from everyone) on this?

AjinkyaDahale commented:
@Pesc0 if it is not done now that is slightly more work to change the includes in cards 4 (and 3, if these changes are done outside of that) whenever the changes are made. If you believe making these changes right now will need more time (because of testing etc.) than you have, I'm OK with delaying, but otherwise I think it's best done here.

Pesc0 commented:
@AjinkyaDahale I had some time today to look more more in depth, here are my observations:
- All the functions you mentioned are static helpers, quite simple to move. 
- At the same time they seem to touch a lot of files, so it would be better to chose wisely: is ElementMap the right place to put them?
- Also they seem to be part of a block of helpers: you didn't mention `hasMissingElement`, `noElementName`, `findElementName`, `hasMappedElementName`. I think the block should be moved all together, not just the ones you mentioned. 

Given all of this, I was reluctant towards this change simply because we are struggling to merge this, so it did not seem a good idea to add more stuff on top of it. Since this seems a simple change, would you be able to submit a PR? I would do it myself but since I don't know the codebase very well it would take quite some time.

AjinkyaDahale commented:
>  Since this seems a simple change, would you be able to submit a PR? I would do it myself but since I don't know the codebase very well it would take quite some time.

@Pesc0, are you asking me to make a PR to your branch or directly to the main?. Also, at this point I believe you know better about this part of the code than I do.

Pesc0 commented:
Alright I've managed to move them to PostfixStringReferences, as it seemed a better place to me (since they are static functions).  I'm thinking of renaming PostfixStringReferences to ElementNamingUtils. Pushing changes soon

Pesc0 commented:
@AjinkyaDahale I did not have to touch `SketchObject` (that you mentioned would drop the complexgeodata dependency). Could you check that these changes actually do what you were thinking of?

AjinkyaDahale commented:
> @AjinkyaDahale I did not have to touch `SketchObject` (that you mentioned would drop the complexgeodata dependency). Could you check that these changes actually do what you were thinking of?

I think there was a miscommunication. ComplexGeoData isn't currently used in SketchObject in master, but it is planned in card 4.

Pesc0 commented:
Ah right, I've missed that. Still, do these changes reflect what you expected?

AjinkyaDahale commented on src/App/ElementNamingUtils.h:
```diff
@@ -0,0 +1,59 @@
+{
+
+/// Special prefix to mark the beginning of a mapped sub-element name
+const std::string ELEMENT_MAP_PREFIX                = ";";
```
Please see https://github.com/FreeCAD/FreeCAD/pull/8923#issuecomment-1476693963 and the link within it :D.

Pesc0 replied:
Ah snap I've missed that. See also the discussion here https://github.com/FreeCAD/FreeCAD/commit/9af8e6a0556bd6def909a2787613ab0af698a06a#commitcomment-113554654.
Is there a better way to do this?

AjinkyaDahale replied:
Not sure of the variable naming convention. I would suggest staying with the rest of FreeCAD there.

Coming to the optimization, Pretty sure that would be asking too much. It could consider that a future file may import and want to use the string related functions, so it'll create the whole object rather than just the string. Perhaps that's why it's a strict no for libraries. This is all thought experiment though.

AjinkyaDahale commented:
@Pesc0 sorry for the late reply. I just checked and to answer your question, it indeed does reflect the changes. Thanks again.

JohnAD commented:
The latest PR I pushed to your `elementmap` branch has a fix for handling the singleton better. Essentially, before, all four tests would pass just fine when you ran one at a time. But if you ran all four of them in a row, the Setup/Teardown would create a run-time error. So, it now does a suite-wide setup for the `Application` singleton, and the individual test `Setup` only creates a new Document and leaves the singleton alone.

Pesc0 commented:
@JohnAD Thanks! I've left a comment there

JohnAD commented:
We will probably want to merge to master soon as this branch and master are starting to diverge quite a bit.

I'd like to add quite a few more "mocked operation" tests to the unit testing. But, frankly, I think that could be done later after the merge.

It would, however, be good to get the remaining basic unit tests done; at _least_ a tiny one per public method (except the save/restore methods):

- [ ] MappedName addName(MappedName& name, const IndexedName& idx, const ElementIDRefs& sids, bool overwrite, IndexedName* existing);
- [ ] MappedName setElementName(const IndexedName& element, const MappedName& name, long masterTag, const ElementIDRefs* sid = nullptr, bool overwrite = false);
- [x] void encodeElementName(char element_type, MappedName& name, std::ostringstream& ss, ElementIDRefs* sids, long masterTag, const char* postfix = 0, long tag = 0, bool forceTag = false) const;
- [ ] MappedName hashElementName(const MappedName& name, ElementIDRefs& sids) const;
- [ ] MappedName dehashElementName(const MappedName& name) const;
- [ ] bool erase(const MappedName& name);
- [ ] bool erase(const IndexedName& idx);
- [x] unsigned long size() const;
- [ ] bool empty() const;
- [ ] IndexedName find(const MappedName& name, ElementIDRefs* sids = nullptr) const;
- [ ] MappedName find(const IndexedName& idx, ElementIDRefs* sids = nullptr) const;
- [ ] std::vector<std::pair<MappedName, ElementIDRefs>> findAll(const IndexedName& idx) const;
- [ ] bool hasChildElementMap() const;
- [ ] void hashChildMaps(long masterTag);
- [ ] void addChildElements(long masterTag, const std::vector<MappedChildElements>& children);
- [ ] std::vector<MappedChildElements> getChildElements() const;
- [ ] std::vector<MappedElement> getAll() const;

The only reason I think to push off the unused save/restore methods is that it really isn't very testable until the next card IMO (if not the next phase overall).

Does this sound like a finishing plan? @chennes ?

Pesc0 commented:
I've moved things around a bit, hopefully this does not mess up what you were working on.
> at least a tiny one per public method

Note that I have made some methods private, since in RT's branch they seemed to be used only internally. See commit https://github.com/FreeCAD/FreeCAD/pull/9175/commits/5730c4e366ee358cb79ad7824046cfd41ef583da. Then seeing you had used `getAll()` in your test I've reverted some of those changes https://github.com/FreeCAD/FreeCAD/pull/9175/commits/61886b4847b00102d9a97e9ebe5a6f4466cfcdba. I think the simpler the interface the better, thats why I moved as much as possible to private. 

> this branch and master are starting to diverge quite a bit

Not true, I have been rebasing frequently ;)

JohnAD commented:
So that we don't accidentally duplicate effort: I would like to write the unit tests for hasChildElementMap, hashChildMaps, addChildElements, and getChildElements next.

Also, I'm increasing not liking the name `masterTag` that I came up with for the current tag parameter. How does `selfTag` or `currentTag` or `ownTag` sound to you?

Pesc0 commented:
> So that we don't accidentally duplicate effort: I would like to write the unit tests for hasChildElementMap, hashChildMaps, addChildElements, and getChildElements next.

Before you do that could you please ensure that in realthunder's branch they are actually used externally? These are the same functions that I touched in the last commit I mentioned, and I'm not sure if they should be private or not. Not that testing them would hurt, but if they are just used internally we should test the public stuff first.

> Also, I'm increasing not liking the name masterTag that I came up with for the current tag parameter. How does selfTag or currentTag or ownTag sound to you?

Could you please describe briefly what exactly does the tag do / mean?

JohnAD commented:
> Could you please describe briefly what exactly does the tag do / mean?

I'm running from the context of the code, so it is possible my understanding is not full or complete. @chennes,  please correct me if wrong.

But it is my understanding that the tag is used to uniquely id an object/face/vector/etc .

In the simple case, it can be ignored mostly. But when a operation is involved, such as a union, this becomes more important.

So, if a cube (tag = 1) and another cube (tag = 2) are "fused" into a new shape (tag = 3); and you follow through the history of the new shape, you can identify the "current tag" (3) is different than the other objects (1 and 2). This helps prevent duplication of the same objects when serializing the toponame.

It's **very** possible that tags are used for other things also.

Complicating things further, elements of an object, such as faces, vectors(lines) and points also have unique tags. Again, it is to distinguish between the items involved in any operation.

JohnAD commented:
- `ComplexGeoData::getMappedChildElements` directly calls `getChildElements`

    -  however nothing calls `getMappedChildElements`. So, you can delete this one and leave a note in card 3's Issue if you would like.

- `ComplexGeoData::hasChildElementMap()` calls ElementMap -> `hasChildElementMap` and that is called by Mod/Part/App/PropertyTopoShape.cpp

- `ComplexGeoData::hashChildMaps()` calls ElementMap -> `hashChildMaps` and that is called by Mod/Part/App/PropertyTopoShape.cpp

- `ComplexGeoData::setMappedChildElements` directly calls `addChildElements`. That is called by `ComplexGeoData::setMappedChildElements`. That is called by Mod/Part/App/PropertyTopoShape.cpp

For these last three, it is possible that the ComplexGeoData methods might be removed if we add a direct ElementMapPtr getter. But that is for the next card to consider.

Pesc0 commented:
Thanks for the explanation, it cleared things up a bit. I'm thinking `elementTag`? or `currentTag`. self / own make me think the tag is referring to the map itself.

As for the functions it looks like they should be public then. Thanks for investigating that.

Pesc0 commented:
Looks like addName, hashElementName, dehashElementName are used only internally as well? Should I make them private?

JohnAD commented:
`addName` .... yes, private. `hashElementName` ... private. `dehashElementName` ... also private.

AjinkyaDahale commented on tests/src/App/ElementMap.cpp:
```diff
@@ -3,16 +3,198 @@
 #include "gtest/gtest.h"
 
 #include "App/ElementMap.h"
+#include <App/Application.h>
```
Is there any reason this and the above headers are imported with different formats?

Pesc0 commented:
When I have a bit of time I should be able to add the remaining tests, not including the ones John is working on of course.

AjinkyaDahale commented:
@Pesc0 at some point you should probably also look into squashing the commits to a reasonable number. For example, all fixes can get merged with the commit they fix.

Pesc0 commented:
Except for a failing test that should be investigated I think this is now pretty much ready.

@AjinkyaDahale I would do that but at this point I dont even remember what I did and for what reason, and I'm really afraid of breaking stuff.

AjinkyaDahale commented:
> @AjinkyaDahale I would do that but at this point I dont even remember what I did and for what reason, and I'm really afraid of breaking stuff.

I would advise creating a new branch from this one and experimenting with it. It's more important if you can't recall your steps. This can tell if you missed something important, for example.

Pesc0 commented:
I've just tried reorganizing and squashing the commits, but it did not work so well. The end result (comparing the branches before and after the operation) is that the files are not the same.

AjinkyaDahale commented:
@Pesc0 thanks for checking it out. @JohnAD do you have any opinion on the squashing?

Pesc0 commented:
I think it's a good idea to reorganize and rename the commits. The difference between branches most likely comes from the fact that I've messed up something.

JohnAD commented:
I generally squash a PR into a single commit; letting the single commit message and the changes themselves express the meaning of the merge.

But if you are wanting to preserve commit history while still doing some squashes...that can get really complicated. Git isn't really designed to handle that easily.

JohnAD commented:
IIRC, squashing to a single commit is an option in GitHub's PR process. So, if you ware willing to have a single commit, write up the wording you want Chris to use when completing the PR.

AjinkyaDahale commented:
@JohnAD I would steer away from squashing into a single commit. If you do choose that, Id suggest keeping the text of all the commits.

chennes commented:
I don't know that there's a lot of value in preserving the commit history here -- my strategy with the Toponaming work has been to make one commit that is as close as possible to just real thunder's original code, and then a subsequent commit (possibly more than one) that is my additions and changes. So if this can be arranged like that then some squashing and cleanup may be in order. Otherwise, let's just squash the whole thing on merge.

chennes commented:
Any progress on that failing `ElementMapTest.eraseIndexedName` test? I obviously don't want to merge until we are passing CI...

Pesc0 commented:
I've written the test intentionally failing because to me that's what makes the most sense. The question is if the elementmap code is correct, not the test. 

I would like to get an ACK from @JohnAD on the other tests I've added as well, just to ensure I didn't write anything completely bogus.

chennes commented:
I can't merge while CI fails -- what's the path forward here?

JohnAD commented:
On first examination, the `erase` is not erasing the indexed name and it's mapping. Instead, it is erasing the first matching type of indexed name. So, erasing "Edge1" will erase "Edge*". So erasing "Edge1" again will erase "Edge*" and if "Edge2" is in the list of indexed names, it will erase that one next.

I'm not done investigating. I'm going to try to "deduce meaning" from usage the original toponaming branch. Guessing, either:

1. it is badly named and should be renamed something like `eraseFirstMatchingNameType`.  (Where "type" means a string pointer to "Edge". I really don't like that `IndexedName` call the string prefix a `type`. But that is a different discussion.) or,
2. this never comes up as an issue in most scenarios, but we should add some extra code to make this work in all scenarios anyway.

I have a suspicion it is scenario 1. But I'll report back soon.

JohnAD commented:
Actually, even as I hit send on the above commend, this can't possibly be correct behavior. I'll do more homework :smile: .

JohnAD commented:
Yeah, it's just a bug. And, it was probably never caught because the `bool` that is returned is never used.

So, I'll going to remove the return value since it adds to value. But, i'll also add to the test to make sure the "same type" does not over-erase. I'll send a quick PR.

John

Pesc0 commented:
Thanks John, I've mirrored the same changes to the other erase function, `erase(const MappedName& name)`

Pesc0 commented:
Managed to squash things a bit. Perhaps last time I was too aggressive with the squashing.

Should be ready to merge, but before that please someone take a look at the tests in particular, let's not make the mistake of merging broken stuff that will bite us in the future.

Pesc0 commented:
does this work? I am on linux so I'll have to wait for CI

JohnAD commented:
All pipeline checks pass! Unfortunately I don't have access to a Windows machine to try running it as a final check.

Pesc0 commented:
CI reports no issues, I think there's no need for you to run it John, thanks for being available though.

chennes commented:
Let's do it, then!

luzpaz commented:
Heads-up, this PR added trailing whitespace

```
Warning: src/App/ElementMap.h:59:    /** Default constructor: hooks internal functions to \c signalSaveDocument and <-- trailing whitespace
Warning: src/App/ElementMap.h:60:     * \c signalStartRestoreDocument. This is related to the save and restore process <-- trailing whitespace
Warning: src/App/ElementMap.h:68:     * <-- trailing whitespace
Warning: src/App/ElementMap.h:76:     * @param s: serialized stream <-- trailing whitespace
Warning: src/App/ElementMap.h:80:    /** Deserialize and restore this map. This function restores \c childMaps and <-- trailing whitespace
Warning: src/App/ElementMap.h:81:     * \c postfixes from the stream, then calls the other (private) restore function with those <-- trailing whitespace
Warning: src/App/ElementMap.h:186:     * @param s: serialized stream <-- trailing whitespace
Warning: src/App/ElementMap.h:208:     * @param existing: out variable: if not overwriting, and \c name is already <-- trailing whitespace
Warning: src/App/ElementMap.h:216:     * if it was not present in the map. <-- trailing whitespace
Warning: src/App/MappedName.cpp:75:        pos = this->rfind(POSTFIX_DECIMAL_TAG); <-- trailing whitespace
Warning: src/App/MappedName.cpp:110:        //                     |              |   |   |  <-- trailing whitespace
```
